<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>Algorithms and framework for Nonnegative Matrix Factorization (NMF).. NMF 0.8</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="description" content="">
    <meta name="author" content="">

    <!-- Le styles -->
    <link href="css/bootstrap.css" rel="stylesheet">
    <style>
      body {padding-top: 60px;}
      h2 {padding-top: 20px}
    </style>
    <link href="css/bootstrap-responsive.css" rel="stylesheet">
    <link href="css/highlight.css" rel="stylesheet">
    <link href="css/staticdocs.css" rel="stylesheet">

    <!-- Le HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>

  <body>

    <!-- START_Navigation bar -->
    <div class="navbar navbar-fixed-top">
  <div class="navbar-inner">
    <div class = "container">
      <a class="brand" href="index.html">NMF 0.8</a>
      <div class="nav">
        <ul class="nav">
       <li><a href="_MAN.html">Documentation</a></li>
       <li><a href="_REFERENCES.html">References</a></li>
       </ul>
   </div>
    </div>
  </div>
</div>
    <!-- END_Navigation bar -->

    <div class="container">
       
       <h1>References</h1><br />

<div class="row">
<div class="span8">
	<ol>
  	<li><p>Lee DD and Seung HS (1999).
&ldquo;Learning the parts of objects by non-negative matrix factorization.&rdquo;
<EM>Nature</EM>, <B>401</B>(6755), pp. 788&ndash;91.
ISSN 0028-0836, <a href="http://dx.doi.org/10.1038/44565">http://dx.doi.org/10.1038/44565</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/10548103">http://www.ncbi.nlm.nih.gov/pubmed/10548103</a>.</li>
  	<li><p>Paatero P and Tapper U (1994).
&ldquo;Positive matrix factorization: A non-negative factor model with optimal utilization of error estimates of data values.&rdquo;
<EM>Environmetrics</EM>, <B>5</B>(2), pp. 111&ndash;126.
<a href="http://dx.doi.org/10.1002/env.3170050203">http://dx.doi.org/10.1002/env.3170050203</a>, <a href="http://www3.interscience.wiley.com/cgi-bin/abstract/113468839/ABSTRACT">http://www3.interscience.wiley.com/cgi-bin/abstract/113468839/ABSTRACT</a>.</li>
  	<li><p>Wang G, Kossenkov AV and Ochs MF (2006).
&ldquo;LS-NMF: a modified non-negative matrix factorization algorithm utilizing uncertainty estimates.&rdquo;
<EM>BMC bioinformatics</EM>, <B>7</B>, pp. 175.
ISSN 1471-2105, <a href="http://dx.doi.org/10.1186/1471-2105-7-175">http://dx.doi.org/10.1186/1471-2105-7-175</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/16569230">http://www.ncbi.nlm.nih.gov/pubmed/16569230</a>.</li>
  	<li><p>Hoyer P (2004).
&ldquo;Non-negative matrix factorization with sparseness constraints.&rdquo;
<EM>The Journal of Machine Learning Research</EM>, <B>5</B>, pp. 1457&ndash;1469.
<a href="http://portal.acm.org/citation.cfm?id=1044709">http://portal.acm.org/citation.cfm?id=1044709</a>.</li>
  	<li><p>Carmona-Saez P, Pascual-Marqui RD, Tirado F, Carazo JM and Pascual-Montano A (2006).
&ldquo;Biclustering of gene expression data by Non-smooth Non-negative Matrix Factorization.&rdquo;
<EM>BMC bioinformatics</EM>, <B>7</B>, pp. 78.
ISSN 1471-2105, <a href="http://dx.doi.org/10.1186/1471-2105-7-78">http://dx.doi.org/10.1186/1471-2105-7-78</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/16503973">http://www.ncbi.nlm.nih.gov/pubmed/16503973</a>.</li>
  	<li><p>Brunet J, Tamayo P, Golub TR and Mesirov JP (2004).
&ldquo;Metagenes and molecular pattern discovery using matrix factorization.&rdquo;
<EM>Proceedings of the National Academy of Sciences of the United States of America</EM>, <B>101</B>(12), pp. 4164&ndash;9.
ISSN 0027-8424, <a href="http://dx.doi.org/10.1073/pnas.0308531101">http://dx.doi.org/10.1073/pnas.0308531101</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/15016911">http://www.ncbi.nlm.nih.gov/pubmed/15016911</a>.</li>
  	<li><p>Pascual-Montano A, Carazo JM, Kochi K, Lehmann D and Pascual-marqui RD (2006).
&ldquo;Nonsmooth nonnegative matrix factorization (nsNMF).&rdquo;
<EM>IEEE Trans. Pattern Anal. Mach. Intell</EM>, <B>28</B>, pp. 403&ndash;415.</li>
  	<li><p>Hutchins LN, Murphy SM, Singh P and Graber JH (2008).
&ldquo;Position-dependent motif characterization using non-negative matrix factorization.&rdquo;
<EM>Bioinformatics (Oxford, England)</EM>, <B>24</B>(23), pp. 2684&ndash;90.
ISSN 1367-4811, <a href="http://dx.doi.org/10.1093/bioinformatics/btn526">http://dx.doi.org/10.1093/bioinformatics/btn526</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/18852176">http://www.ncbi.nlm.nih.gov/pubmed/18852176</a>.</li>
  	<li><p>Badea L (2008).
&ldquo;Extracting gene expression profiles common to colon and pancreatic adenocarcinoma using simultaneous nonnegative matrix factorization.&rdquo;
<EM>Pacific Symposium on Biocomputing. Pacific Symposium on Biocomputing</EM>, <B>290</B>, pp. 267&ndash;78.
ISSN 1793-5091, <a href="http://www.ncbi.nlm.nih.gov/pubmed/18229692">http://www.ncbi.nlm.nih.gov/pubmed/18229692</a>.</li>
  	<li><p>Kim H and Park H (2007).
&ldquo;Sparse non-negative matrix factorizations via alternating non-negativity-constrained least squares for microarray data analysis.&rdquo;
<EM>Bioinformatics (Oxford, England)</EM>, <B>23</B>(12), pp. 1495&ndash;502.
ISSN 1460-2059, <a href="http://dx.doi.org/10.1093/bioinformatics/btm134">http://dx.doi.org/10.1093/bioinformatics/btm134</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/17483501">http://www.ncbi.nlm.nih.gov/pubmed/17483501</a>.</li>
  	<li><p>Kim H and Park H (2007).
&ldquo;Sparse non-negative matrix factorizations via alternating non-negativity-constrained least squares for microarray data analysis.&rdquo;
<EM>Bioinformatics (Oxford, England)</EM>, <B>23</B>(12), pp. 1495&ndash;502.
ISSN 1460-2059, <a href="http://dx.doi.org/10.1093/bioinformatics/btm134">http://dx.doi.org/10.1093/bioinformatics/btm134</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/17483501">http://www.ncbi.nlm.nih.gov/pubmed/17483501</a>.</li>
  	<li><p>Kim H and Park H (2007).
&ldquo;Sparse non-negative matrix factorizations via alternating non-negativity-constrained least squares for microarray data analysis.&rdquo;
<EM>Bioinformatics (Oxford, England)</EM>, <B>23</B>(12), pp. 1495&ndash;502.
ISSN 1460-2059, <a href="http://dx.doi.org/10.1093/bioinformatics/btm134">http://dx.doi.org/10.1093/bioinformatics/btm134</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/17483501">http://www.ncbi.nlm.nih.gov/pubmed/17483501</a>.</li>
  	<li><p>Kim H and Park H (2007).
&ldquo;Sparse non-negative matrix factorizations via alternating non-negativity-constrained least squares for microarray data analysis.&rdquo;
<EM>Bioinformatics (Oxford, England)</EM>, <B>23</B>(12), pp. 1495&ndash;502.
ISSN 1460-2059, <a href="http://dx.doi.org/10.1093/bioinformatics/btm134">http://dx.doi.org/10.1093/bioinformatics/btm134</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/17483501">http://www.ncbi.nlm.nih.gov/pubmed/17483501</a>.</li>
  	<li><p>Kim H and Park H (2007).
&ldquo;Sparse non-negative matrix factorizations via alternating non-negativity-constrained least squares for microarray data analysis.&rdquo;
<EM>Bioinformatics (Oxford, England)</EM>, <B>23</B>(12), pp. 1495&ndash;502.
ISSN 1460-2059, <a href="http://dx.doi.org/10.1093/bioinformatics/btm134">http://dx.doi.org/10.1093/bioinformatics/btm134</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/17483501">http://www.ncbi.nlm.nih.gov/pubmed/17483501</a>.</li>
  	<li><p>Lee DD and Seung H (2001).
&ldquo;Algorithms for non-negative matrix factorization.&rdquo;
<EM>Advances in neural information processing systems</EM>.
<a href="http://scholar.google.com/scholar?q=intitle:Algorithms+for+non-negative+matrix+factorization\#0">http://scholar.google.com/scholar?q=intitle:Algorithms+for+non-negative+matrix+factorization\#0</a>.</li>
  	<li><p>Van Benthem M and Keenan MR (2004).
&ldquo;Fast algorithm for the solution of large-scale non-negativity-constrained least squares problems.&rdquo;
<EM>Journal of Chemometrics</EM>, <B>18</B>(10), pp. 441&ndash;450.
ISSN 0886-9383, <a href="http://dx.doi.org/10.1002/cem.889">http://dx.doi.org/10.1002/cem.889</a>, <a href="http://doi.wiley.com/10.1002/cem.889">http://doi.wiley.com/10.1002/cem.889</a>.</li>
  	<li><p>Van Benthem M and Keenan MR (2004).
&ldquo;Fast algorithm for the solution of large-scale non-negativity-constrained least squares problems.&rdquo;
<EM>Journal of Chemometrics</EM>, <B>18</B>(10), pp. 441&ndash;450.
ISSN 0886-9383, <a href="http://dx.doi.org/10.1002/cem.889">http://dx.doi.org/10.1002/cem.889</a>, <a href="http://doi.wiley.com/10.1002/cem.889">http://doi.wiley.com/10.1002/cem.889</a>.</li>
  	<li><p>Boutsidis C and Gallopoulos E (2008).
&ldquo;SVD based initialization: A head start for nonnegative matrix factorization.&rdquo;
<EM>Pattern Recognition</EM>, <B>41</B>(4), pp. 1350&ndash;1362.
ISSN 00313203, <a href="http://dx.doi.org/10.1016/j.patcog.2007.09.010">http://dx.doi.org/10.1016/j.patcog.2007.09.010</a>, <a href="http://linkinghub.elsevier.com/retrieve/pii/S0031320307004359">http://linkinghub.elsevier.com/retrieve/pii/S0031320307004359</a>.</li>
  	<li><p>Golub TR, Slonim DK, Tamayo P, Huard C, Gaasenbeek M, Mesirov JP, Coller H, Loh ML, Downing JR, Caligiuri Ma, Bloomfield CD and Lander ES (1999).
&ldquo;Molecular classification of cancer: class discovery and class prediction by gene expression monitoring.&rdquo;
<EM>Science (New York, N.Y.)</EM>, <B>286</B>(5439), pp. 531&ndash;7.
ISSN 0036-8075, <a href="http://www.ncbi.nlm.nih.gov/pubmed/10521349">http://www.ncbi.nlm.nih.gov/pubmed/10521349</a>.</li>
  </ol>
</div>
</div>
          
      <footer class="footer">
        <p class="pull-right"><a href="#">Back to top</a></p>
        <p>Built by <a href="https://github.com/hadley/staticdocs">staticdocs</a>. Styled with <a href="http://twitter.github.com/bootstrap">bootstrap</a>.</p>
      </footer>
    </div> <!-- /container -->
  </body>
</html>