<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
<title>. NMF 0.13</title>
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="author" content="">

<link href="css/bootstrap.css" rel="stylesheet">
<link href="css/bootstrap-responsive.css" rel="stylesheet">
<link href="css/highlight.css" rel="stylesheet">
<link href="css/staticdocs.css" rel="stylesheet">

<!-- d3 word cloud -->
<script src="js/d3.v2.min.js"></script>
<script src="js/d3.layout.cloud.js"></script>

<!--[if lt IE 9]>
  <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="navbar">
  <div class="navbar-inner">
    <div class = "container">
      <a class="brand" href="index.html">NMF 0.13</a>
      <div class="nav">
        <ul class="nav">
        <li><a href=""><i class="icon-home icon-white"></i> Index</a></li>
       <li><a href="_MAN.html">Documentation</a></li>
       <li><a href="_DEMOS.html">Demos</a></li>
       <li><a href="_REFERENCES.html">References</a></li>
       </ul>
   	  </div>
    </div>
  </div>
</div>

    <div class="container">
      <header>
        
      </header>
      
      <h1>References</h1><br />

<div class="row"><div class="span8">
  <ol>
   		<li><p>R Development Core Team (2011).
<EM>R: A Language and Environment for Statistical Computing</EM>.
R Foundation for Statistical Computing, Vienna, Austria.
<a href="http://www.r-project.org">http://www.r-project.org</a>.</li>
   		<li><p>Gentleman RC, Carey VJ, Bates DM, Bolstad B, Dettling M, Dudoit S, Ellis B, Gautier L, Ge Y, Gentry J, Hornik K, Hothorn T, Huber W, Iacus S, Irizarry R, Leisch F, Li C, Maechler M, Rossini AJ, Sawitzki G, Smith C, Smyth G, Tierney L, Yang JYH and Zhang J (2004).
&ldquo;Bioconductor: open software development for computational biology and bioinformatics.&rdquo;
<EM>Genome biology</EM>, <B>5</B>(10), pp. R80.
ISSN 1465-6914, <a href="http://dx.doi.org/10.1186/gb-2004-5-10-r80">http://dx.doi.org/10.1186/gb-2004-5-10-r80</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/15461798">http://www.ncbi.nlm.nih.gov/pubmed/15461798</a>.</li>
   		<li><p>Lee DD and Seung H (2001).
&ldquo;Algorithms for non-negative matrix factorization.&rdquo;
<EM>Advances in neural information processing systems</EM>.
<a href="http://scholar.google.com/scholar?q=intitle:Algorithms+for+non-negative+matrix+factorization\#0">http://scholar.google.com/scholar?q=intitle:Algorithms+for+non-negative+matrix+factorization\#0</a>.</li>
   		<li><p>Li SZ and Hou X (2001).
&ldquo;Learning Spatially Localized, Parts-Based Representation.&rdquo;
<EM>Convergence</EM>, <B>00</B>(C), pp. 1&ndash;6.</li>
   		<li><p>Badea L (2008).
&ldquo;Extracting gene expression profiles common to colon and pancreatic adenocarcinoma using simultaneous nonnegative matrix factorization.&rdquo;
<EM>Pacific Symposium on Biocomputing. Pacific Symposium on Biocomputing</EM>, <B>290</B>, pp. 267&ndash;78.
ISSN 1793-5091, <a href="http://www.ncbi.nlm.nih.gov/pubmed/18229692">http://www.ncbi.nlm.nih.gov/pubmed/18229692</a>.</li>
   		<li><p>Zhang J, Wei L, Feng X, Ma Z and Wang Y (2008).
&ldquo;Pattern expression nonnegative matrix factorization: algorithm and applications to blind source separation.&rdquo;
<EM>Computational intelligence and neuroscience</EM>, <B>2008</B>, pp. 168769.
ISSN 1687-5265, <a href="http://dx.doi.org/10.1155/2008/168769">http://dx.doi.org/10.1155/2008/168769</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/18566689">http://www.ncbi.nlm.nih.gov/pubmed/18566689</a>.</li>
   		<li><p>Kim H and Park H (2007).
&ldquo;Sparse non-negative matrix factorizations via alternating non-negativity-constrained least squares for microarray data analysis.&rdquo;
<EM>Bioinformatics (Oxford, England)</EM>, <B>23</B>(12), pp. 1495&ndash;502.
ISSN 1460-2059, <a href="http://dx.doi.org/10.1093/bioinformatics/btm134">http://dx.doi.org/10.1093/bioinformatics/btm134</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/17483501">http://www.ncbi.nlm.nih.gov/pubmed/17483501</a>.</li>
   		<li><p>Albright R, Cox J, Duling D, Langville AN and Meyer C (2006).
&ldquo;Algorithms, initializations, and convergence for the nonnegative matrix factorization.&rdquo;
Technical Report 919, NCSU Technical Report Math 81706. http://meyer. math. ncsu. edu/Meyer/Abstracts/Publications. html.
<a href="http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.89.2161\&amp;amp;rep=rep1\&amp;amp;type=pdf http://meyer.math.ncsu.edu/Meyer/PS\_Files/NMFInitAlgConv.pdf">http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.89.2161\&amp;amp;rep=rep1\&amp;amp;type=pdf http://meyer.math.ncsu.edu/Meyer/PS\_Files/NMFInitAlgConv.pdf</a>.</li>
   		<li><p>Boutsidis C and Gallopoulos E (2008).
&ldquo;SVD based initialization: A head start for nonnegative matrix factorization.&rdquo;
<EM>Pattern Recognition</EM>, <B>41</B>(4), pp. 1350&ndash;1362.
ISSN 00313203, <a href="http://dx.doi.org/10.1016/j.patcog.2007.09.010">http://dx.doi.org/10.1016/j.patcog.2007.09.010</a>, <a href="http://linkinghub.elsevier.com/retrieve/pii/S0031320307004359">http://linkinghub.elsevier.com/retrieve/pii/S0031320307004359</a>.</li>
   		<li><p>L'Ecuyer P, Simard R and Chen E (2002).
&ldquo;An object-oriented random-number package with many long streams and substreams.&rdquo;
<EM>Operations Research</EM>, <B>50</B>(6), pp. 1073&ndash;1075.
<a href="http://www.jstor.org/stable/3088626">http://www.jstor.org/stable/3088626</a>.</li>
   		<li><p>Hutchins LN, Murphy SM, Singh P and Graber JH (2008).
&ldquo;Position-dependent motif characterization using non-negative matrix factorization.&rdquo;
<EM>Bioinformatics (Oxford, England)</EM>, <B>24</B>(23), pp. 2684&ndash;90.
ISSN 1367-4811, <a href="http://dx.doi.org/10.1093/bioinformatics/btn526">http://dx.doi.org/10.1093/bioinformatics/btn526</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/18852176">http://www.ncbi.nlm.nih.gov/pubmed/18852176</a>.</li>
   		<li><p>Frigyesi A and HÃ¶glund M (2008).
&ldquo;Non-negative matrix factorization for the analysis of complex gene expression data: identification of clinically relevant tumor subtypes.&rdquo;
<EM>Cancer informatics</EM>, <B>6</B>(2003), pp. 275&ndash;92.
ISSN 1176-9351, <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2623306/">http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2623306/</a>.</li>
   		<li><p>Brunet J, Tamayo P, Golub TR and Mesirov JP (2004).
&ldquo;Metagenes and molecular pattern discovery using matrix factorization.&rdquo;
<EM>Proceedings of the National Academy of Sciences of the United States of America</EM>, <B>101</B>(12), pp. 4164&ndash;9.
ISSN 0027-8424, <a href="http://dx.doi.org/10.1073/pnas.0308531101">http://dx.doi.org/10.1073/pnas.0308531101</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/15016911">http://www.ncbi.nlm.nih.gov/pubmed/15016911</a>.</li>
   		<li><p>Pascual-Montano A, Carazo JM, Kochi K, Lehmann D and Pascual-marqui RD (2006).
&ldquo;Nonsmooth nonnegative matrix factorization (nsNMF).&rdquo;
<EM>IEEE Trans. Pattern Anal. Mach. Intell</EM>, <B>28</B>, pp. 403&ndash;415.</li>
   		<li><p>Lee DD and Seung HS (1999).
&ldquo;Learning the parts of objects by non-negative matrix factorization.&rdquo;
<EM>Nature</EM>, <B>401</B>(6755), pp. 788&ndash;91.
ISSN 0028-0836, <a href="http://dx.doi.org/10.1038/44565">http://dx.doi.org/10.1038/44565</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/10548103">http://www.ncbi.nlm.nih.gov/pubmed/10548103</a>.</li>
   		<li><p>Paatero P and Tapper U (1994).
&ldquo;Positive matrix factorization: A non-negative factor model with optimal utilization of error estimates of data values.&rdquo;
<EM>Environmetrics</EM>, <B>5</B>(2), pp. 111&ndash;126.
<a href="http://dx.doi.org/10.1002/env.3170050203">http://dx.doi.org/10.1002/env.3170050203</a>, <a href="http://www3.interscience.wiley.com/cgi-bin/abstract/113468839/ABSTRACT">http://www3.interscience.wiley.com/cgi-bin/abstract/113468839/ABSTRACT</a>.</li>
   		<li><p>Hoyer P (2004).
&ldquo;Non-negative matrix factorization with sparseness constraints.&rdquo;
<EM>The Journal of Machine Learning Research</EM>, <B>5</B>, pp. 1457&ndash;1469.
<a href="http://portal.acm.org/citation.cfm?id=1044709">http://portal.acm.org/citation.cfm?id=1044709</a>.</li>
   		<li><p>Carmona-Saez P, Pascual-Marqui RD, Tirado F, Carazo JM and Pascual-Montano A (2006).
&ldquo;Biclustering of gene expression data by Non-smooth Non-negative Matrix Factorization.&rdquo;
<EM>BMC bioinformatics</EM>, <B>7</B>, pp. 78.
ISSN 1471-2105, <a href="http://dx.doi.org/10.1186/1471-2105-7-78">http://dx.doi.org/10.1186/1471-2105-7-78</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/16503973">http://www.ncbi.nlm.nih.gov/pubmed/16503973</a>.</li>
   		<li><p>Wang G, Kossenkov AV and Ochs MF (2006).
&ldquo;LS-NMF: a modified non-negative matrix factorization algorithm utilizing uncertainty estimates.&rdquo;
<EM>BMC bioinformatics</EM>, <B>7</B>, pp. 175.
ISSN 1471-2105, <a href="http://dx.doi.org/10.1186/1471-2105-7-175">http://dx.doi.org/10.1186/1471-2105-7-175</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/16569230">http://www.ncbi.nlm.nih.gov/pubmed/16569230</a>.</li>
   		<li><p>Van Benthem M and Keenan MR (2004).
&ldquo;Fast algorithm for the solution of large-scale non-negativity-constrained least squares problems.&rdquo;
<EM>Journal of Chemometrics</EM>, <B>18</B>(10), pp. 441&ndash;450.
ISSN 0886-9383, <a href="http://dx.doi.org/10.1002/cem.889">http://dx.doi.org/10.1002/cem.889</a>, <a href="http://doi.wiley.com/10.1002/cem.889">http://doi.wiley.com/10.1002/cem.889</a>.</li>
   		<li><p>Golub TR, Slonim DK, Tamayo P, Huard C, Gaasenbeek M, Mesirov JP, Coller H, Loh ML, Downing JR, Caligiuri Ma, Bloomfield CD and Lander ES (1999).
&ldquo;Molecular classification of cancer: class discovery and class prediction by gene expression monitoring.&rdquo;
<EM>Science (New York, N.Y.)</EM>, <B>286</B>(5439), pp. 531&ndash;7.
ISSN 0036-8075, <a href="http://www.ncbi.nlm.nih.gov/pubmed/10521349">http://www.ncbi.nlm.nih.gov/pubmed/10521349</a>.</li>
   		<li><p>Cichocki A, Zdunek R and Amari S (2008).
&ldquo;Nonnegative matrix and tensor factorization.&rdquo;
<EM>IEEE Signal Processing Magazine</EM>, <B>25</B>, pp. 142&ndash;145.</li>
   		<li><p>Berry M, Browne M, Langville AN, Pauca V and Plemmons R (2007).
&ldquo;Algorithms and applications for approximate nonnegative matrix factorization.&rdquo;
<EM>Computational Statistics \&amp; Data Analysis</EM>, <B>52</B>(1), pp. 155&ndash;173.
<a href="http://www.sciencedirect.com/science/article/pii/S0167947306004191">http://www.sciencedirect.com/science/article/pii/S0167947306004191</a>.</li>
   		<li><p>Chu M, Diele F, Plemmons R and Ragni S (2004).
&ldquo;Optimality, computation, and interpretation of nonnegative matrix factorizations.&rdquo;
<EM>SIAM Journal on Matrix Analysis</EM>, pp. 4&ndash;8030.
<a href="http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.61.5758">http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.61.5758</a>.</li>
   		<li><p>Gao Y and Church G (2005).
&ldquo;Improving molecular cancer class discovery through sparse non-negative matrix factorization.&rdquo;
<EM>Bioinformatics (Oxford, England)</EM>, <B>21</B>(21), pp. 3970&ndash;5.
ISSN 1367-4803, <a href="http://dx.doi.org/10.1093/bioinformatics/bti653">http://dx.doi.org/10.1093/bioinformatics/bti653</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/16244221">http://www.ncbi.nlm.nih.gov/pubmed/16244221</a>.</li>
   		<li><p>Roux JL and e A (2008).
&ldquo;Adaptive template matching with shift-invariant semi-NMF.&rdquo;
<EM>Science And Technology</EM>.
<a href="http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.143.6846&amp;rep=rep1&amp;type=pdf">http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.143.6846&amp;rep=rep1&amp;type=pdf</a>.</li>
   		<li><p>Ding C, Li T and Jordan MI (2010).
&ldquo;Convex and semi-nonnegative matrix factorizations.&rdquo;
<EM>IEEE transactions on pattern analysis and machine intelligence</EM>, <B>32</B>(1), pp. 45&ndash;55.
ISSN 1939-3539, <a href="http://dx.doi.org/10.1109/TPAMI.2008.277">http://dx.doi.org/10.1109/TPAMI.2008.277</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/19926898">http://www.ncbi.nlm.nih.gov/pubmed/19926898</a>.</li>
  </ol>
</div></div>
      
      <footer>
      <p class="pull-right"><a href="#">Back to top</a></p>
<p>Built by <a href="https://github.com/hadley/staticdocs">staticdocs</a>. Styled with <a href="http://twitter.github.com/bootstrap">bootstrap</a>.</p>
      </footer>
    </div>
  </body>
</html>